lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexFuture.java

```diff
@@ -34,37 +36,89 @@
   // Otherwise tasklet has not completed.
   private Optional<TOutput> userResult = null;
   private Exception userException;
+  private boolean cancelled = false;
   private final CountDownLatch countDownLatch = new CountDownLatch(1);
   private final ThreadPoolStage<TOutput> stage;
+  private final VortexMaster vortexMaster;
+  private final int taskletId;
 
   /**
    * Creates a {@link VortexFuture}.
    */
-  public VortexFuture() {
+  @Private
+  public VortexFuture(final VortexMaster vortexMaster, final int taskletId) {
+    this.vortexMaster = vortexMaster;
+    this.taskletId = taskletId;
     stage = null;
   }
 
   /**
    * Creates a {@link VortexFuture} with a callback.
    */
-  public VortexFuture(final EventHandler<TOutput> callbackHandler) {
+  @Private
+  public VortexFuture(final VortexMaster vortexMaster,
+                      final int taskletId,
+                      final EventHandler<TOutput> callbackHandler) {
+    this.vortexMaster = vortexMaster;
+    this.taskletId = taskletId;
     stage = new ThreadPoolStage<>(callbackHandler, 1);
   }
 
   /**
-   * TODO[REEF-502]: Support Vortex Tasklet(s) cancellation by user.
+   * Sends a cancel signal and blocks and waits until the task is cancelled, completed, or failed.
+   * @return true if task did not start or was cancelled, false if task failed or completed
    */
   @Override
   public boolean cancel(final boolean mayInterruptIfRunning) {
-    throw new UnsupportedOperationException("Cancel not yet supported");
+    try {
+      return cancel(mayInterruptIfRunning, Optional.<Long>empty(), Optional.<TimeUnit>empty());
+    } catch (final TimeoutException e) {
+      // This should never happen.
```

Comment:
If this should never happen, then what about leaving a log message in `WARNING` level?